File: LeetCode/1114. Print in Order/solution.py
# class Foo:
#     def __init__(self):
#         self.a, self.b, self.c = False, False, False


#     def first(self, printFirst: 'Callable[[], None]') -> None:
        
#         # printFirst() outputs "first". Do not change or remove this line.
#         printFirst()
#         self.a = True


#     def second(self, printSecond: 'Callable[[], None]') -> None:
        
#         # printSecond() outputs "second". Do not change or remove this line.
#         while not self.a:
#             pass
#         printSecond()
#         self.b = True


#     def third(self, printThird: 'Callable[[], None]') -> None:
        
#         # printThird() outputs "third". Do not change or remove this line.
#         while not self.b:
#             pass
#         printThird()
#         self.c = True
        
        
# from threading import Lock

# class Foo:
#     def __init__(self):
#         self.locks = (Lock(),Lock())
#         self.locks[0].acquire()
#         self.locks[1].acquire()
        
#     def first(self, printFirst):
#         printFirst()
#         self.locks[0].release()
        
#     def second(self, printSecond):
#         with self.locks[0]:
#             printSecond()
#             self.locks[1].release()
            
            
#     def third(self, printThird):
#         with self.locks[1]:
#             printThird()
            

# from threading import Event

# class Foo:
#     def __init__(self):
#         self.done = (Event(),Event())
        
#     def first(self, printFirst):
#         printFirst()
#         self.done[0].set()
        
#     def second(self, printSecond):
#         self.done[0].wait()
#         printSecond()
#         self.done[1].set()
            
#     def third(self, printThird):
#         self.done[1].wait()
#         printThird()

        
from threading import Semaphore

class Foo:
    def __init__(self):
        self.gates = (Semaphore(0),Semaphore(0))
        
    def first(self, printFirst):
        printFirst()
        self.gates[0].release()
        
    def second(self, printSecond):
        with self.gates[0]:
            printSecond()
            self.gates[1].release()
            
    def third(self, printThird):
        with self.gates[1]:
            printThird()

File: LeetCode/1114. Print in Order/solution.cpp
class Foo {
private:
    bool turnFirst, turnSecond, turnThird;
    condition_variable cv;
    mutex my_mutex;
public:
    Foo()
        : turnFirst(true), turnSecond(false), turnThird(false)
    {
        
    }

    void first(function<void()> printFirst) {
        unique_lock<mutex> lk(my_mutex);
        cv.wait(lk, [this](){
            return turnFirst;
        });
        // printFirst() outputs "first". Do not change or remove this line.
        printFirst();
        turnSecond = true;
        turnFirst = false;
        cv.notify_all();
    }

    void second(function<void()> printSecond) {
        unique_lock<mutex> lk(my_mutex);
        cv.wait(lk, [this](){
            return turnSecond;
        });
        // printSecond() outputs "second". Do not change or remove this line.
        printSecond();
        turnThird = true;
        turnSecond = false;
        cv.notify_all();
    }

    void third(function<void()> printThird) {
        unique_lock<mutex> lk(my_mutex);
        cv.wait(lk, [this](){
            return turnThird;
        });
        // printThird() outputs "third". Do not change or remove this line.
        printThird();
        turnFirst = true;
        turnThird = false;
        cv.notify_all();
    }
};













#include <semaphore.h>

class Foo {

protected:
    sem_t firstJobDone;
    sem_t secondJobDone;

public:

    Foo() {
        sem_init(&firstJobDone, 0, 0);
        sem_init(&secondJobDone, 0, 0);
    }

    void first(function<void()> printFirst) {
        // printFirst() outputs "first".
        printFirst();
        sem_post(&firstJobDone);
    }

    void second(function<void()> printSecond) {
        sem_wait(&firstJobDone);
        // printSecond() outputs "second".
        printSecond();
        sem_post(&secondJobDone);
        
    }

    void third(function<void()> printThird) {
        sem_wait(&secondJobDone);
        // printThird() outputs "third".
        printThird();
    }
};


// 作者：LeetCode
// 链接：https://leetcode-cn.com/problems/print-in-order/solution/an-xu-da-yin-by-leetcode/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。